runtime/gc_vlhgc/VLHGCAccessBarrier.cpp

```diff
@@ -294,17 +274,17 @@ MM_VLHGCAccessBarrier::jniGetPrimitiveArrayCritical(J9VMThread* vmThread, jarray
 		copyArrayCritical(vmThread, indexableObjectModel, functions, &data, arrayObject, isCopy);
 	} else if (!indexableObjectModel->isInlineContiguousArraylet(arrayObject)) {
 		/* an array having discontiguous extents is another reason to force the critical section to be a copy */
+		bool isAllIndexableDataContiguousEnabled = indexableObjectModel->isVirtualLargeObjectHeapEnabled();
 #if defined(J9VM_GC_ENABLE_DOUBLE_MAP)
-		MM_EnvironmentVLHGC *env = MM_EnvironmentVLHGC::getEnvironment(vmThread);
-		if (indexableObjectModel->isDoubleMappingEnabled()) {
+		isAllIndexableDataContiguousEnabled = isAllIndexableDataContiguousEnabled || indexableObjectModel->isDoubleMappingEnabled();
+#endif /* defined(J9VM_GC_ENABLE_DOUBLE_MAP) */
+		if (isAllIndexableDataContiguousEnabled) {
+#if defined(J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+			MM_EnvironmentVLHGC *env = MM_EnvironmentVLHGC::getEnvironment(vmThread);
 			fj9object_t *arrayoidPtr = indexableObjectModel->getArrayoidPointer(arrayObject);
 			if (indexableObjectModel->isArrayletDataDiscontiguous(arrayObject)) {
```

Comment:
It's confusing to use Discontiguous variant of an API, if we are already in the path where isAllIndexableDataContiguousEnabled is true. Not sure what was the intention here - just to recognize non-0 sized array. Perhaps we should use non-adjacent or size > 0 type of checks.